### PR TITLE
Chores/implement redirect for users 

### PR DESCRIPTION
**Pull request by Alexanderlacuna**
This PR add these supports:
* [] redirect users to login page if the service requires login
* [] after successful login redirect users to the correct page.
See this for the full issue description:
https://issues.genenetwork.org/issues/gn-auth/implement-redirect-on-login

### CODE REVIEW

File: gn2/wqflask/oauth2/checks.py [hunk 1]
```diff
@@ -12,29 +12,41 @@
...
+from .request_utils import authserver_authorise_uri
 
 
 def require_oauth2(func):
     """Decorator for ensuring user is logged in."""
     @wraps(func)
     def __token_valid__(*args, **kwargs):
         """Check that the user is logged in and their token is valid."""
-
         def __clear_session__(_no_token):
```
Comment (jnduli) [inline gn2/wqflask/oauth2/checks.py]: drop this function?

File: gn2/wqflask/oauth2/checks.py [hunk 2]
```diff
@@ -12,29 +12,41 @@
...
 
+        def __redirect_to_login__(_token):
+            """
+            Save the current user request to session then
+            redirect to the login page.
+            """
+            if request.method == "GET":
+                redirect_url = url_for(request.endpoint, **request.args)
+            else:
+                redirect_url = "/"
```
Comment (jnduli) [inline gn2/wqflask/oauth2/checks.py:32]: A minor improvement would be to assume all requests have a corresponding `GET`, so we could change this to:
```
# Note: this isn't tested but writen to demonstrate the idea
try:
  redirect_url = url_for(rquest.endpoint, _method="GET", **request.args)
except BuildError:
  redirect_url = "/"
```

Here's the tests I've done for this:
```
(Pdb) url_for("gsearch_updating", _method="GET")
*** werkzeug.routing.BuildError: Could not build url for endpoint 'gsearch_updating' ('GET'). Did you mean to use methods ['OPTIONS', 'POST']?
(Pdb) url_for("gsearch_updating", _method="POST")
'/gsearch_updating'
```

The advantage is that for endpoints that support both `GET` and `POST`, we'll redirect to the page that had the form

Comment (Alexanderlacuna) [reply gn2/wqflask/oauth2/checks.py:32]: Thanks, made the requested changes.

**Review by jnduli — APPROVED** [review]

File: gn2/wqflask/oauth2/session.py [hunk 3]
```diff
@@ -118,3 +119,10 @@ def toggle_token_refreshing():
 def is_token_refreshing():
     """Returns whether the token is being refreshed or not."""
     return session_info().get("token_refreshing", False)
+
+def set_redirect_url(url):
```
Comment (fredmanglis) [inline gn2/wqflask/oauth2/session.py:123]: This looks good, but you should probably add a related `clear_redirect_url` function to clean up after yourself once the login is complete, but just before you do the redirect.

File: gn2/wqflask/oauth2/toplevel.py [hunk 4]
```diff
@@ -81,7 +81,8 @@ def __success__(token):
                 "token": session.user_token(),
                 "logged_in": True
             })
-            return redirect("/")
+            redirect_url = session.session_info().get("redirect_url", "/")
+            return redirect(redirect_url)
```
Comment (fredmanglis) [inline gn2/wqflask/oauth2/toplevel.py:85]: Clear the "redirect_url" value in the session here, just before the `return redirect(redirect_url)` statement. Don't leave the now unnecessary value in place.

**Review by fredmanglis — COMMENTED** [review]
Other than the 2 comments regarding clearing the temporary value from the session, this looks good.

**Comment by Alexanderlacuna** [comment]
@fredmanglis Thanks for the review, Well noted.

**Pull request merged**